### DISCUSSION TITLE
Is there any example using uSwift instead of the Swift standard library?

### DISCUSSION THREAD

**Discussion by kkebo** [compnerd/uswift#18] [General]
I can build uSwift itself by following README.md, but do you have any examples of projects using uSwift?

**Comment by compnerd** [comment]
Unfortunately, nothing public.  However, the C++ interop work was experimenting with this library as a means for enabling testing on various platforms.  Is there something specific that you are looking for (e.g. how to actually use to this instead of the regular standard library)?

**Reply by kkebo** [reply]
@compnerd Sorry for the long long delay. I wanted to use uSwift instead of the standard library to write an OS, just like `no_std` is used to write an OS in Rust. Do you expect such a use case?

**Reply by MaxDesiatov** [reply]
Sorry for interjecting, there's an example of this in [Hleeb repository](https://github.com/MaxDesiatov/Hleeb), where I'm integrating uSwift as a subdirectory via a symlink to a sibling directory. But it probably could be integrated as a git submodule, or with some CMake command that checks it out for you.

Unfortunately, uSwift by itself is far from ready for this purpose at the moment. The main blocker is the lack of pointer types support.

**Reply by compnerd** [reply]
@kkk669 yes, that is something that the project is intended to support (and I did experiment with doing that to pretty good success) though I did end up with some workarounds as I hadn't done much work on the pointers yet at @MaxDesiatov points out :).

**Reply by kkebo** [reply]
Thank you both! I'm going to start by experimenting with an example @MaxDesiatov taught me. I'm looking forward to uSwift's maturity (and if there is anything I can do to help, I'll be happy to contribute).

**Reply by MaxDesiatov** [reply]
I think [linking issue with `MemoryLayout`](https://github.com/compnerd/uswift/pull/38) is a big know roadblock on our path towards pointers support. It seems to be Windows-specific, I'm not having issues with it when cross-compiling to bare metal ARM for example. As for the rest of types and functions that pointers depend on, I'll get back to those at some point I hope, but in the meantime everything is up for grabs, right?

**Reply by compnerd** [reply]
@kkk669 - patches to improve uSwift are welcome (@MaxDesiatov can attest to that).  I definitely would love to improve the size tracking and have more analysis into where the space is going.  so that we can understand how to best keep the library reasonable for embedded systems.
